### ISSUE TITLE
Early Stopping/Patience Epochs

### ISSUE THREAD

**Issue by FarhAnonymous** [ArgoHA/custom_d_fine#2]
I need help figuring out on early stopping. I was going through the original repo and even there isn't anything like this. Can you please help me putting patience epochs and which file?

**Comment by ArgoHA** [comment]
Early stopping is not implemented. In this repo I use OneCyclerLR scheduler, so it's usually good to wait till the end of training (and set a realistic number of epochs in the first place).
What is your use case? I can add early stopping in the next update.

**Comment by FarhAnonymous** [comment]
Thank you for your response. I'm training on palm oil tree detection datasets and comparing multiple object detection models (including D-FINE) for my research. The challenge is that I can't determine an exact 'realistic' number of epochs in advance that would be fair for all models in the comparison.

My concern is that without early stopping, I have to either:
a) Risk training for too few epochs where some models don't reach convergence
b) Train for too many epochs and potentially give some models an unfair advantage due to overfitting

Early stopping would ensure all models are compared at their optimal performance point regardless of how many epochs it takes. If you could add this feature or help me add it to the original repo of D-Fine?

**Comment by ArgoHA** [comment]
I can add that feature to this repo today, but integrating to the original repo takes too much time unfortunately. Maybe I will do it later too.
At this moment I can recommend this:
Try 60 epochs and "m" model and check valid metrics graph. If you see plateouing - you can stop training manually. If you see scores are still rising - you need more epochs. Even if you don't stop - you have the best model saved (based on valid mAP and F1). So you should not see overfitting.

Meanwhile I will add early stopping to this repo. Will let you know when it's ready.


![Image](https://github.com/user-attachments/assets/58f95f6f-5df3-46ae-abeb-8aea332ce73f)

**Comment by FarhAnonymous** [comment]
Will wait for it to get updated then or you can just guide me on what edits I can make in the original repo to set early stoppage cus I can't do this trial and error as I have limited compute resources on collab and I have to train the model on all the depths n, s, m, l, x.

Thank you again!

**Comment by ArgoHA** [comment]
@FarhAnonymous early stopping implemented, check out the README and config.yaml and let me know if you need any help
`early_stopping: 10`

**Comment by FarhAnonymous** [comment]
Thanks for the update, will play with it and let you know if I need more help again.

**Comment by ArgoHA** [comment]
Here is a quick example how to train D-FINE with colab

https://colab.research.google.com/drive/1ZV12qnUQMpC0g3j-0G-tYhmmdM98a41X?usp=sharing

**Comment by FarhAnonymous** [comment]
I'm impressed with your customized D-FINE implementation and appreciate your help with the early stopping functionality. Beyond that specific feature, could you elaborate on the key architectural or implementation differences between your version and the original D-FINE? I'm particularly interested in whether there are any advantages in terms of:

Model efficiency (reduced parameters, inference speed optimizations)
Training performance (convergence speed, final accuracy)
Additional functionality (beyond early stopping)
Memory optimization techniques

**Comment by ArgoHA** [comment]
No changes on the model architecture or loss functions. Everything else is custom, here are some advantages:
- Mosaic augmentation and some other augs that are absent in original repo
- More metrics (I implemented some metrics in the original repo too)
- After training is done - runs a test to calculate the optimal conf threshold
- Batch accumulation
- Keep ratio of the image and use paddings or use simple resize (in the original repo only simple resize is available)
- Visualisation of preprocessed images, model predictions and ground truth
- Unified configuration file for all scrips
- Early stopping
- Export to ONNX, OpenVino, TensorRT
- Inference code for TensorRT, PyTorch, OpenVINO
- OneCyclerLR

At least with my datasets I see a small improvement in the accuracy, mainly because of a scheduler and augmentations. Everything else was done for convenience (checking model preds during eval process, checking postprocessed images, one config file, inference code...)

**Comment by FarhAnonymous** [comment]
I have created my dataset using Roboflow and it's already in coco format and split into train, val, and test. So, I am skipping the preprocessing and splitting scripts and If my dataset is already augmented, should I disable some of the built-in augmentations? Which ones are critical to keep? (Multiple augmentations can potentially over-distort images)

**Comment by ArgoHA** [comment]
1. this repo uses YOLO format, so you need to get `images and `labels` folders first (with all of your data)
2. preprocessing is for image conversions. If you have everything in jpgs, you can skip it.
3. Splitting is for train/val/test splits. You need to apply it after 
4. Usually augmentations are applied on the fly, but if you have your specific augs done offline, you can try only using mosaic and setting other augs to 0

There are a lot of tools to convert coco2yolo, including roboflow

**Comment by ArgoHA** [comment]
Closing this issue, as early stopping is now implemented.
